Implement uniform sampling of bits from field elements

Vindaar opened the pull request:
Ref: https://github.com/Plonky3/Plonky3/issues/613 and https://github.com/tcoratger/whir-p3/issues/278

### Overview

While looking at the issues for WHIR, https://github.com/tcoratger/whir-p3/issues/278 caught my eye. It seemed like we could achieve a decent middle ground between performance and uniform sampling. My detailed thoughts are here: https://github.com/tcoratger/whir-p3/issues/278#issuecomment-3303801269

### Main idea

Essentially, the idea to address the bias in the random sampling of bits is to make use of the fact that for most small fields (BabyBear, KoalaBear and Goldilocks at least), we can actually sample a lot of bits uniformly if we:
- accept a 1/P probability of a panic 
- or perform rejection sampling in very rare cases (1/P)

Due to the field structure of these 3 fields:
- $p_{BB} = 2^{31} - 2^{27} + 1$
- $p_{KB} = 2^{31} - 2^{24} + 1$
- $p_G = 2^{64} - 2^{32} + 1$

there is a natural set of bits (27, 24, 32), which up to 1 field element have exactly the same number of zeroes and ones. In other words, if we only sample from $[0, p-2]$, we get uniform bits. For uniformly sampled field elements this implies a 1/P chance of hitting $p-1$, which would be invalid. In that case, either panic or resample.

More generally, we can sample $k$ bits uniformly, if we perform rejection sampling on $[0, m_k)$ with

```math
m_k = ⌊P / 2^k⌋ · 2^k
```

elements (resample if $x \geq m_k$), as that interval has exactly a multiple of $2^k$ elements.

I've added a feature `uniform-sampling-may-panic`, which if activated will panic instead of performing rejection sampling. The idea is for folks worried about recursion costs, it may be preferable to just accept the 1/P chance of a proof failure.

### Summary

At the moment I leave this as a draft PR. For a few reasons:
- because I still haven't written any kind of tests (nor finalized a prototype implementation in WHIR, only local testing). 
- I don't know if this solution in general is of interest in Plonky3
- I'll be out of office until Sep 30 now.

I also don't know if the feature flag is a good idea or not. I'm especially unsure about the fact that the feature is used in two crates (challenger & mersenne31). At the moment I think if one sets the feature differently between the two crates, one ends up with a case where one circumvents the compile time error I added for `Mersenne31`. 
Also, I specifically left the original `sample_bits` untouched for now.

**Note**: If anyone has good ideas on how to cut down on the duplication for the `UniformSamplingField` implementations, I'm all ears.

tcoratger commented on challenger/src/duplex_challenger.rs:
```diff
@@ -196,11 +196,86 @@ where
+
+        // If this feature is enabled, we accept a small chance O(1/P) that we
+        // panic in this function, namely if the sampled field element `result >= m`.
+        #[cfg(feature = "uniform-sampling-may-panic")]
```
Yes, that's not usually what we do in the codebase. The only thing we use this for is to enable or not parallelism: for this we have a parallel cfg. But for the rest we don't normally use cfg.

tcoratger commented on challenger/src/grinding_challenger.rs:
```diff
@@ -66,6 +99,47 @@ where
+        assert!(self.check_witness_uniform_may_panic(bits, witness));
+        witness
+    }
+}
```
Another proposition for your consideration to refactor the code

```suggestion
impl<F, P, const WIDTH: usize, const RATE: usize> UniformGrindingChallenger
    for DuplexChallenger<F, P, WIDTH, RATE>
where
    F: UniformSamplingField + PrimeChief64,
    P: CryptographicPermutation<[F; WIDTH]>,
{
    #[instrument(name = "grind uniform for proof-of-work witness", skip_all)]
    fn grind_uniform(&mut self, bits: usize) -> Self::Witness {
        // Call the generic grinder with the "resample" checking logic.
        self.grind_generic(bits, |challenger, witness| {
            challenger.check_witness_uniform(bits, witness)
        })
    }

    #[instrument(name = "grind uniform may panic for proof-of-work witness", skip_all)]
    fn grind_uniform_may_panic(&mut self, bits: usize) -> Self::Witness {
        // Call the generic grinder with the "panic" checking logic.
        self.grind_generic(bits, |challenger, witness| {
            challenger.check_witness_uniform_may_panic(bits, witness)
        })
    }
}

impl<F, P, const WIDTH: usize, const RATE: usize> DuplexChallenger<F, P, WIDTH, RATE>
where
    F: PrimeChief64,
    P: CryptographicPermutation<[F; WIDTH]>,
{
    /// A generic, private helper for PoW grinding, parameterized by the checking function.
    fn grind_generic<CHECK>(&mut self, bits: usize, mut check_fn: CHECK) -> F
    where
        CHECK: FnMut(&mut Self, F) -> bool + Sync + Send,
    {
        // Maybe check that bits is greater than 0?
        assert!(bits < (usize::BITS as usize), "bit count must be valid");
        assert!((1u64 << bits) < F::ORDER_U64, "bit count exceeds field order");

        // The core parallel brute-force search logic.
        let witness = (0..F::ORDER_U64)
            .into_par_iter()
            .map(|i| unsafe {
                // This is safe as i is always in range.
                F::from_canonical_unchecked(i)
            })
            .find_any(|&witness| check_fn(&mut self.clone(), witness))
            .expect("failed to find proof-of-work witness");

        // Run the check one last time on the *original* challenger to update its state
        // and confirm the witness is valid.
        assert!(check_fn(self, witness));

        witness
    }
}
```

tcoratger reviewed (approved):
Looks good to me, I think you just need to sort the cargo file to make the CI green :)

What do you think @SyxtonPrime about this?

Vindaar commented on challenger/src/duplex_challenger.rs:
```diff
@@ -204,6 +205,141 @@ where
+    /// triggering a panic / requiring a resample.
+    const MAX_SINGLE_SAMPLE_BITS: usize;
+    /// An array storing the largest value `m_k` for each `k` in [0, 31], such that `m_k`
+    /// is a multiple of `2^k`. `m_k` is defined as:
```
Yeah, adding the less than `P`.

Vindaar commented on challenger/src/duplex_challenger.rs:
```diff
@@ -204,6 +205,141 @@ where
+{
+    /// Generic implementation for uniform bit sampling, parameterized by a strategy.
+    #[inline]
+    fn sample_uniform_bits_with_strategy<S>(&mut self, bits: usize, _strategy: S) -> usize
```
@tcoratger came up with the strategy approach. I initially wrote the code as in https://github.com/Plonky3/Plonky3/pull/1050/commits/8c3473c1ee43fa4febd7dbb6e0647b519bebc7e9. I just kept the argument, as I naively assumed it was required for the generic, but of course we can just call the function with an explicit generic argument instead. Changed.

Vindaar commented:
I've addressed most of your points. Only haven't made changes for the `check_witness` call in `grind` yet and haven't moved the `UniformSamplingField` impl of the Mont31 to the monty crate yet.

tcoratger commented on challenger/src/duplex_challenger.rs:
```diff
@@ -204,6 +208,172 @@ where
+    ///
+    /// \( m_k = ⌊P / 2^k⌋ · 2^k \)
+    ///
+    /// This is used as a rejection sampling threshold (or panic trigger) in `sampling_uniform_bits`, when
```
```suggestion
    /// This is used as a rejection sampling threshold (or error trigger) in `sampling_uniform_bits`, when
```

tcoratger reviewed (commented):
I believe there are still instances of the word "panic" around the PR even though we're now reporting errors: function namings, object naming, documentation, etc.

Vindaar commented on challenger/src/lib.rs:
```diff
@@ -70,6 +70,26 @@ pub trait CanSampleBits<T> {
     fn sample_bits(&mut self, bits: usize) -> T;
 }
 
+pub trait CanSampleUniformBits<F> {
```
Done.

Vindaar commented on challenger/src/grinding_challenger.rs:
```diff
@@ -77,7 +183,10 @@ where
+        assert!(
+            bits < (usize::BITS as usize) && bits > 0,
+            "bit count must be valid"
+        );
```
Whoops, overlooked that one.

SyxtonPrime commented on challenger/src/duplex_challenger.rs:
```diff
@@ -204,6 +205,141 @@ where
+{
+    const MAX_SINGLE_SAMPLE_BITS: usize = MP::MAX_SINGLE_SAMPLE_BITS;
+    const SAMPLING_BITS_M: [u64; 64] = MP::SAMPLING_BITS_M;
+}
```
Hmm I definitely see your point but it's a little strange having this here but the M31 and Goldilocks impls in the respective crates. 

I'd either move them all here, or put them all in the field crates. Happy either way.

SyxtonPrime commented on challenger/src/grinding_challenger.rs:
```diff
@@ -40,6 +43,51 @@ pub trait GrindingChallenger:
+    fn check_witness_uniform_may_panic(&mut self, bits: usize, witness: Self::Witness) -> bool {
+        self.observe(witness);
+        self.sample_uniform_bits_may_panic(bits) == 0
+    }
```
Just a quick reminder to change the name of this

SyxtonPrime reviewed (approved):
LGTM

One minor reminder of a thing to rename before we merge. (Also fix clippy :) )

Vindaar commented:
@SyxtonPrime Not sure where you found that remaining `may_panic`, but that should be gone already. Also just addressed the Clippy errors.

tcoratger commented:
@Vindaar Missing cargo fmt and then we are good to merge
